### PR TITLE
Use fsspec to implement table_names with cloud storage support

### PR DESCRIPTION
**Pull request by gsilvestrin**
See `python/tests/test_io.py` for an example of usage with s3

### CODE REVIEW

File: python/lancedb/util.py [hunk 1]
```diff
@@ -41,3 +41,11 @@ def get_uri_scheme(uri: str) -> str:
         # So we add special handling here for schemes that are a single character
         scheme = "file"
     return scheme
+
+
+# A function that removes the suffix of a string
+# This helper exist because String.removesuffix is only available for Python > 3.9
+def removesuffix(string, suffix):
```
Comment (changhiskhan) [inline python/lancedb/util.py:48]: does `os.path.splitext` work?

File: python/lancedb/db.py [hunk 2]
```diff
@@ -47,11 +49,26 @@ def table_names(self) -> list[str]:
...
+        if scheme == "file":
+            filesystem = fs.LocalFileSystem()
+        elif scheme == "s3":
+            filesystem = fs.S3FileSystem()
+        elif scheme == "gs":
+            filesystem = fs.GcsFileSystem()
+        else:
+            raise NotImplementedError(
+                "Unsupported scheme: " + scheme
+            )
```
Comment (wjones127) [inline python/lancedb/db.py:63]: We can let PyArrow get the path for us if we want:

```suggestion
        try:
            filesystem, path = fs.FileSystem.from_uri(self.uri)
        except pa.ArrowInvalid:
            raise NotImplementedError(
                "Unsupported scheme: " + self.uri
            )
```

**Review by wjones127 — APPROVED** [review]
Seems good. Had a comment where we might simplify, but given this is temporary I think it's fine to merge as is.

**Review by changhiskhan — APPROVED** [review]

File: python/pyproject.toml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 [project]
 name = "lancedb"
 version = "0.1.4"
-dependencies = ["pylance>=0.4.17", "ratelimiter", "retry", "tqdm"]
+dependencies = ["pylance>=0.4.17", "ratelimiter", "retry", "tqdm", "fsspec", "s3fs", "gcsfs"]
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/pyproject.toml]: Isn't pyarrow already support s3 and gcs  natively?

**Pull request merged**